hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseCompactor.java

```diff
@@ -31,16 +33,30 @@
 
   private static final long serialVersionUID = 1L;
 
+  protected final transient Object writeClientUpdateLock = new Object();
+  protected final transient List<BaseHoodieWriteClient<T, I, K, O>> oldCompactionClientList = new ArrayList<>();
+
   protected transient BaseHoodieWriteClient<T, I, K, O> compactionClient;
 
+  protected boolean isCompactionRunning = false;
+
   public BaseCompactor(BaseHoodieWriteClient<T, I, K, O> compactionClient) {
     this.compactionClient = compactionClient;
   }
 
   public abstract void compact(HoodieInstant instant) throws IOException;
 
   public void updateWriteClient(BaseHoodieWriteClient<T, I, K, O> writeClient) {
-    this.compactionClient = writeClient;
+    synchronized (writeClientUpdateLock) {
+      if (!isCompactionRunning) {
+        this.compactionClient.close();
+      } else {
+        // Store the old compaction client so that they can be closed
```

Comment:
can't we just tie the life cycle of the write client to the life cycle of a single execution of clustering or compaction. i.e from DeltaSync level, when we submit the runnable for a executing a given compaction or clustering, we make a new write client. is that not sufficient? simpler?

Reply:
Agree very strongly with the point above: unless there are very strong argument why we can not re-init Async Service itself, i believe we should follow t/h with an invariant that AS and Write Client lifecycles are coupled and solve this concurrency control issues at the root

Reply:
Immutability is hard but very powerful property we should lean in on it as much as possible, only detour from it when there's no other choice (mostly for perf reasons)